cedar-policy-core/src/authorizer/partial_response.rs

```diff
@@ -0,0 +1,481 @@
+
+    /// Returns the set of [`PolicyID`]s that were definitely satisfied
+    pub fn definitely_satisfied(&self) -> impl Iterator<Item = &PolicyID> {
+        self.satisfied_permit_ids()
+            .chain(self.satisfied_forbid_ids())
+    }
+
+    /// Returns the set of [`PolicyID`]s that encountered errors
+    pub fn definitely_errored(&self) -> impl Iterator<Item = &PolicyID> {
+        self.false_permits
+            .iter()
+            .chain(self.false_forbids.iter())
+            .filter_map(did_error)
+    }
+
+    /// Returns an over-approximation of the set of determining policies
+    pub fn may_be_determining(&self) -> impl Iterator<Item = &PolicyID> {
+        if self.satisfied_forbids.is_empty() {
+            // We have no definitely true forbids, so the over approx is everything that is true or potentially true
+            Either::Left(
+                self.satisfied_permit_ids()
+                    .chain(self.residual_permits.keys())
+                    .chain(self.residual_forbids.keys()),
+            )
+        } else {
+            // We have definitely true forbids, so we know only things that can determine is
+            // true forbids and potentially true forbids
+            Either::Right(
+                self.satisfied_forbid_ids()
+                    .chain(self.residual_forbids.keys()),
+            )
+        }
+    }
+
+    /// Returns an under-approximation of the set of determining policies.
+    pub fn definitely_determining(&self) -> impl Iterator<Item = &PolicyID> {
+        // If there are no true forbids or potentially true forbids,
+        // the the under approximation is the true permits
+        if self.satisfied_forbids.is_empty() && self.residual_forbids.is_empty() {
+            Either::Left(self.satisfied_permit_ids())
+        } else {
+            // Otherwise it's the true forbids
+            Either::Right(self.satisfied_forbid_ids())
+        }
+    }
+
+    /// Returns the set of non-trivial (meaning more than just `true` or `false`) residuals expressions
+    pub fn nontrivial_residuals(&self) -> impl Iterator<Item = &Expr> {
+        self.residual_forbids
+            .values()
+            .chain(self.residual_permits.values())
+            .map(|(r, _)| r)
+    }
+
+    /// Returns every policy as a residual expression
+    pub fn all_residuals(&self) -> impl Iterator<Item = &Expr> {
+        self.all_permit_residuals()
+            .chain(self.all_forbid_residuals())
+            .map(|(_, (residual, _))| residual)
+    }
+
+    /// Returns all residuals expressions that come from [`Effect::Permit`] policies
+    fn all_permit_residuals(&self) -> impl Iterator<Item = (&PolicyID, (&Expr, &Annotations))> {
+        let trues = self
+            .satisfied_permits
+            .iter()
+            .map(|(id, a)| (id, (&self.true_expr, a)));
+        let falses = self
+            .false_permits
+            .iter()
+            .map(|(id, (_, a))| (id, (&self.false_expr, a)));
```

Comment:
for sake of discussion: Should this method (and `all_forbid_residuals()` below) actually include the `false` residuals?  or should we only include the `true` and nontrivial residuals?